### DISCUSSION TITLE
Issues and Questions about Obstacle Distance, Simulator and Trajectory Plot

### DISCUSSION THREAD

**Discussion by MayDGT** [skhatiri/UAV-Testing-Competition#2] [Platform] [answered]
Hi! I have several questions about the rules and the tools employed in this competition:
* During generation, do we have access to the plan file? I.e., can we rely on the knowledge of the origin, destination, and  waypoints?
* What does the curve with “original” label represent in the generated results?
* How is the budget computed?

Thanks for your answers in advance!

**Comment by MayDGT** [comment]
Moreover, I have some other questions about the simulation of the UAV:
* We produced an example in which the obstacle is not too tall (h=5) and the UAV flies over the obstacle without any collision. However, the code returns that the minimum distance with the obstacle is 0.
![19-10-15-13-29-ASUS](https://github.com/skhatiri/UAV-Testing-Competition/assets/38600921/bd5f6957-e05e-4625-8cd3-966be46b22d1)
* We observed that, if we give an invalid test with two overlapping objects, the simulation does not tell us that it is invalid, but it produces a weird path for the UAV. How can we know if a test is invalid?
![20-10-15-18-48-ASUS](https://github.com/skhatiri/UAV-Testing-Competition/assets/38600921/676cc876-f143-4828-8b32-045824e659d1)
* We produced a test in which the minimum distance is 0, but the UAV reaches the destination. The figure shows that the drone goes through the building.
![20-10-15-56-44-ASUS](https://github.com/skhatiri/UAV-Testing-Competition/assets/38600921/d366c5b1-de31-45ea-9e06-e651cc76d4a1)

Thanks in advance!

**Reply by skhatiri** [reply]
Regarding the remaining questions: 

> We produced an example in which the obstacle is not too tall (h=5) and the UAV flies over the obstacle without any collision. However, the code returns that the minimum distance with the obstacle is 0.

- Thanks for spotting this. The current implementation of the **distance** only considers the (x,y) dimensions. The reason is that we expected the obstacles to be tall enough that the drone had no way to go above them. 
- We will update the platform to also consider this in the next few days. I will mention it here when it is done. 

> We observed that, if we give an invalid test with two overlapping objects, the simulation does not tell us that it is invalid, but it produces a weird path for the UAV. How can we know if a test is invalid?
- The simulator behaves non-deterministically when the obstacles are overlapping since they can not be physically placed in the provided position. They may fall on each other, or the drone, or be unstable. These test cases are considered invalid. 
- The platform does not internally check the validity of the placements, but you need to do this and make sure you are generating valid test cases. 
- you can check the intersection of the obstacles using [this method](https://github.com/skhatiri/Aerialist/blob/11899ab40e23fc8dad4690a1bb5ff2601b9f4d70/aerialist/px4/obstacle.py#L95C1-L96C56), or develop you own module for these kinds of checks. 

> We produced a test in which the minimum distance is 0, but the UAV reaches the destination. The figure shows that the drone goes through the building.
- It seems that probably the drone flies over one of the obstacles, and because of the **distance** issue mentioned above, you get the 0 distance. 
- Let us know if you have the same issue after we update the distance function

Thanks for the good questions.
We look forward to your submission.

**Reply by skhatiri** [reply]
update: 

- a temporary fix for the distance method is now available with the recent commits to Aerialist ([this](https://github.com/skhatiri/Aerialist/commit/43b1e31488981ff5ad511dd457e3c90c37daae12 )) and the competition code samples ([this](https://github.com/skhatiri/UAV-Testing-Competition/commit/bf657f58db3b6f6cee1ae45ecdcc7bebedb8a516)) 
- It only considers the Z dimension while the drone is flying over the obstacle. it is still ignored when it is away from it. 
- You should pull the latest version both from Aerialist and the competition platform

Let us know if you keep having the above issues.

**Reply by MayDGT** [reply]
Hi, regarding the distance calculation, we find another issue:

Obstacles:
```
size1 = Obstacle.Size(l=17.0, w=18.7, h=25)
position1 = Obstacle.Position(x=1.5, y=27.3, z=0, r=69.8)

size2 = Obstacle.Size(l=3.8, w=7.2, h=25)
position2 = Obstacle.Position(x=-12.2, y=34.5, z=0, r=66.1)

size3 = Obstacle.Size(l=5.5, w=3.2, h=25)
position3 = Obstacle.Position(x=-18.8, y=29.1, z=0, r=50.8)
```
Output:
```
entry - INFO - test finished...
entry - INFO - LOG:/src/aerialist/results/09-11-19-42-41-c6265.ulg
entry - ERROR - program terminated:Multi-part geometries do not provide a coordinate sequence
Traceback (most recent call last):
  File "aerialist/entry.py", line 296, in main
    args.func(args)
  File "aerialist/entry.py", line 227, in run_experiment
    DroneTest.plot(test, test_results)
  File "aerialist/px4/drone_test.py", line 140, in plot
    Trajectory.plot_multiple(
  File "aerialist/px4/trajectory.py", line 192, in plot_multiple
    distance = ave_trajectory.min_distance_to_obstacles(obstacles)
  File "aerialist/px4/trajectory.py", line 341, in min_distance_to_obstacles
    return min(obst.distance(line) for obst in obstacles)
  File "aerialist/px4/trajectory.py", line 341, in <genexpr>
    return min(obst.distance(line) for obst in obstacles)
  File "aerialist/px4/obstacle.py", line 104, in distance
    min_z = min(p[2] for p in intersection.coords)
  File "/usr/local/lib/python3.8/dist-packages/shapely/geometry/base.py", line 958, in coords
    raise NotImplementedError(
NotImplementedError: Multi-part geometries do not provide a coordinate sequence
```

**Reply by japeltom** [reply]
I confirm that we get the same error with the same test scenario.

**Reply by skhatiri** [reply]
Hi @MayDGT @japeltom ,

Thanks for spotting this. 
Apparently, in your test scenarios, the drone trajectory has multiple overlaps with the obstacle locations. 
This needed further handling, which is available now with [this commit](https://github.com/skhatiri/Aerialist/commit/d5e27db3c649f688abc7504d268abf7c04979141). 

- make sure to pull the latest from Dockerhub and GitHub

let me know if the issue is fixed

**Comment by skhatiri** [comment, marked as answer]
Hi, Thanks for the questions.
>     During generation, do we have access to the plan file? I.e., can we rely on the knowledge of the origin, destination, and  waypoints?
- yes, you have access to the case study (test description yaml file), as well as all the files references in it, including the mission plan, parameters, commands, etc. 
 
>     What does the curve with the “original” label represent in the generated results?
- if you provide a reference "ulg" file in the 'assertion' property of the test cases, you will have the flight trajectory of the given flight plotted as the "original" flight.  This only matters if you wish to compare visually the current flight with a previous one. 

>     How is the budget computed?
- the budget is the total number of simulations you are allowed to do during the test generation. 
- Each individual simulation executed during the test generation is calculated once. If you run the same test case n times - n simulations - (to check for a nondeterministic behavior), it will be calculated n times. 
- We will put a limit on this total number to limit the cost of the test generation process. 
- Of course, the less you use from the assigned budget, the better.

**Comment by MayDGT** [comment]
> > We produced a test in which the minimum distance is 0, but the UAV reaches the destination. The figure shows that the drone goes through the building.
> 
> * It seems that probably the drone flies over one of the obstacles, and because of the **distance** issue mentioned above, you get the 0 distance.
> * Let us know if you have the same issue after we update the distance function

Hi, Thank you for the answers! We have tried the latest code and the **distance** calculation is correct now. 
Regarding this question, the heights of the three obstacles were set as 20. As shown in the new figure, the minimum distance is 0, but the drone will go through the top obstacle and reach the destination. So is this the mechanism of the simulation?
![27-10-09-16-31-ASUS](https://github.com/skhatiri/UAV-Testing-Competition/assets/38600921/7d6b9f36-1d69-4129-abb7-486365aabd99)

**Comment by Sne4kers** [comment]
Hi! I have the same issue. Plots are made on old version, so don't mind the distances. 
![27-10-07-10-43-vmi14](https://github.com/skhatiri/UAV-Testing-Competition/assets/40452269/178a822a-599b-4c32-90fe-a09b11407dd3)
![24-10-15-53-58-vmi14](https://github.com/skhatiri/UAV-Testing-Competition/assets/40452269/ac6ad84b-c0ff-46da-b816-4b7fc601aa3d)
It happens in different scenarios, despite each obstacle's height being 200 and the drone flies as if they are never there.

**Comment by skhatiri** [comment]
@MayDGT @Sne4kers 
Thank you both for reporting this issue. 
The problem was that the 3rd and 4th obstacles were not handled correctly when using Docker simulation and they were ignored by the simulator. It should be fixed now with [this commit](https://github.com/skhatiri/Aerialist/commit/e0a2afeae18602c7549d5a55a770c1ae9659b841).
![30-10-15-04-12-9410](https://github.com/skhatiri/UAV-Testing-Competition/assets/16767722/87c7c9b5-f518-4b17-bd01-f77073f4f3f1)
make sure to pull the latest version of [Aerialist](https://github.com/skhatiri/Aerialist) and its [docker Image](https://hub.docker.com/repository/docker/skhatiri/aerialist/)

@Sne4kers too tall obstacles (200m) could get physically unstable, and potentially end up falling down (which you won't see in the plots). I suggest keeping the height around 20 or so to be sure this does not happen. 

Let me know if you keep having any of these issues. 
you can post your test description file to help me with the debugging.

**Reply by japeltom** [reply]
I think it would be helpful if somebody updated the main readme file to include pointers to these issues. Now everybody must go through all these posts to find essential information. Also: why not just make it a rule that all obstacles have height 20?

**Reply by skhatiri** [reply]
Thanks for the suggestion, I will summarise the important notes in the documentation as well. 

Regarding the obstacle height, we do not want to restrict the participants. Playing with the height could also be a way to generate challenging test scenarios.

**Comment by MayDGT** [comment]
Some other questions:
* Does the UAV have a size (i.e., we need to consider its occupancy) or we can consider it as a point without size?
* Regarding invalid tests
    * How do we know if the flight mission is physically feasible? Are there some constraints on the distance between objects?
    * How do we know if a generated test is invalid? Is waiting for the timeout the only way?
* Do we need to use the class _TestCase_ in our generator or can we implement our own?
* How is the budget managed? Do we need to set a maximum number of possible simulations inside our code (as in the random generator given as an example)? Or will the pipeline of the competition stop our generator after a maximum number of simulations have been called?      

For the second question, we generated three obstacles:
```
size1 = Obstacle.Size(l=5.8, w=25.0, h=20)
position1 = Obstacle.Position(x=1.2, y=23.0, z=0, r=40.6)

size2 = Obstacle.Size(l=2.4, w=5.7, h=20)
position2 = Obstacle.Position(x=-12.4, y=31.5, z=0, r=5.7)

size3 = Obstacle.Size(l=4.0, w=10.5, h=20)
position3 = Obstacle.Position(x=-13.9, y=22.9, z=0, r=0.0)
```
and got this error during the execution:
```
Traceback (most recent call last):
  File ".../UAV-Testing-Competition/snippets/test.py", line 23, in <module>
    test.execute()
  File ".../UAV-Testing-Competition/snippets/testcase.py", line 17, in execute
    self.trajectory = self.test_results[0].record
IndexError: list index out of range
```

Thanks for your answers in advance!

**Reply by japeltom** [reply]
Thanks for writing this. We have though about exactly the same matters.

**Reply by Sne4kers** [reply]
The same crashes sometimes used to happen on the old version. From what I understand, the simulation crashes and doesn't produce the flight log. Therefore, you get the error, as self.test_results is simply empty. I have wrapped it in the try/catch and I would assume it's the issue on the side of the simulator.

On the other hand, the new version  produces another error for me:
```
entry - INFO - setting up the test environment...
px4.simulator - INFO - Avoidance is ready (waiting 5 seconds to wrap up)
px4.drone - INFO - Conected to the drone
px4.drone - INFO - updating drone parameters
px4.drone - INFO - uploading mission /io/mission1.plan
entry - INFO - running the test...
px4.drone - INFO - **SCHEDULER STARTED** running 1 commands 
px4.drone - INFO - starting mission
px4.drone - INFO - mission started
px4.simulator - INFO - landing detected: starting the timeout
px4.file_helper - ERROR - Error occurred while copying file.
entry - INFO - test finished...
entry - INFO - LOG:None
entry - ERROR - program terminated:'NoneType' object has no attribute 'to_data_frame'
Traceback (most recent call last):
  File "aerialist/entry.py", line 282, in main
    args.func(args)
  File "aerialist/entry.py", line 213, in run_experiment
    DroneTest.plot(test, test_results)
  File "aerialist/px4/drone_test.py", line 128, in plot
    Trajectory.plot_multiple(
  File "aerialist/px4/trajectory.py", line 115, in plot_multiple
    data_frame = trajectories[i].to_data_frame()
AttributeError: 'NoneType' object has no attribute 'to_data_frame'
```

**Reply by skhatiri** [reply]
@MayDGT Thanks for your questions.

> Does the UAV have a size (i.e., we need to consider its occupancy) or we can consider it as a point without size?
- The UAV used in the simulations is called [**Iris**](http://www.arducopter.co.uk/iris-quadcopter-uav.html) (https://docs.px4.io/v1.12/en/simulation/gazebo.html). You can find its physical characteristics. The distances calculated by the platform consider it as a point (in the center). You can also consider its size in your test generation process if you prefer, but there is no obligation. 
> How do we know if the flight mission is physically feasible? Are there some constraints on the distance between objects?
> How do we know if a generated test is invalid? Is waiting for the timeout the only way?
- Currently, there is no automated way of predicting the feasibility of the test cases. If the simulation takes too long (time out), **especially if more than once**, you can conclude that the simulation did not finish correctly, and probably the timeout passed. 
- You can add some simple sanity checks to the obstacle configurations (e.g., having no overlaps)
> Do we need to use the class TestCase in our generator or can we implement our own?
- This is only a sample code to help you. You can implement your own class or extend it as you wish, but you should keep using the Aerialist functionalities and built-in classes (specifically [DroneTest](https://github.com/skhatiri/Aerialist/blob/master/aerialist/px4/drone_test.py)) for test case properties. 
> How is the budget managed? Do we need to set a maximum number of possible simulations inside our code (as in the random generator given as an example)? Or will the pipeline of the competition stop our generator after a maximum number of simulations have been called?
- The [random generator](https://github.com/skhatiri/UAV-Testing-Competition/blob/master/snippets/random_generator.py) is a good example of how the interface of your final test generator should look like. You need to take the case study, and the generation budget as inputs, and stop the generation process before running out of budget. The platform will not check the budget automatically, but we will make sure that your code does not go over the budget. 
- We will document the submission criteria, and the required interface with more details in the next few days.

**Reply by skhatiri** [reply]
@MayDGT @Sne4kers 
Regarding the errors you mentioned, as you assumed correctly, these kinds of errors are due to the simulation process not finishing as expected. This is either due to some internal errors happening during the simulation and not finishing correctly, or time out expiring. You can simply ignore the test cases that do not come with any `test_results` and assume them as invalid tests with a simple if, or try/except block. 
```
if len(self.test_results)>0:
```

**Comment by tada0s** [comment]
Would it be possible also to get more information on what metrics will be used to evaluate the algorithms:
* What will the total budget be set as?
* How is diversity measured?
* How is simplicity measured? Is it a multiplicative, where less obstacles gives a higher overall score/ranking?

**Reply by skhatiri** [reply]
Thanks for the questions. 
> What will the total budget be set as?
- We will try to give a large budget to the generators, but it depends on our workload for the evaluation phase. We will do our best to give a minimum of 100 simulations to the test generators, but we may be able to go as high as 1000 or more, depending on the total number of participants.
> How is diversity measured?
- In general, the more different failing/challenging test cases you find with the budget, the better. the test cases are considered different if the placement of obstacles are not too identical(e.g., if the two settings differ only in one of the 3 obstacles being moved by only 2cm).
> How is simplicity measured? Is it a multiplicative, where less obstacles gives a higher overall score/ranking?
- The number of obstacles is the most important factor in the simplicity criteria. the fewer obstacles in the scenario, the better.


In general, We are working on documenting the evaluation criteria and the submission requirements in more details in the next few days.
